### ISSUE TITLE
default_constructed_unit_structs sugg causes error

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#12654]
### Summary

.

### Reproducer

I tried this code:

```rust
struct G;

impl PartialEq<B> for G {
	fn eq(&self, _: &B) -> bool {
		true
	}
}

#[derive(Debug, Default)]
struct B;


fn main() {
    let _ = G::eq(&G, dbg!(&<_>::default()));
}

```

Instead, this happened:
```
warning: use of `default` to create a unit struct
  --> ./tests/ui/lint/non-local-defs/generics.rs:14:31
   |
14 |     let _ = G::eq(&G, dbg!(&<_>::default()));
   |                               ^^^^^^^^^^^^ help: remove this call to `default`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_constructed_unit_structs
   = note: requested on the command line with `-W clippy::default-constructed-unit-structs`
```

This does not work lol
```
error: expected one of `>` or `as`, found `<eof>`
   --> src/main.rs:14:30
    |
14  |     let _ = G::eq(&G, dbg!(&<_));
    |                              ^ expected one of `>` or `as`
    |
```

### Version

```text
rustc 1.79.0-nightly (ab5bda1aa 2024-04-08)
binary: rustc
commit-hash: ab5bda1aa70f707014e2e691e43bc37a8819252a
commit-date: 2024-04-08
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.3
```
